pkg/config/config.go

```diff
@@ -0,0 +1,207 @@
+//
+// Copyright 2021 The AFF Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	cm "knative.dev/pkg/configmap"
+)
+
+type Config struct {
+	Storage StorageConfigs
+}
+
+// StorageConfigs contains the configuration to instantiate different storage providers
+type StorageConfigs struct {
+	GCS          GCSStorageConfig
+	OCI          OCIStorageConfig
+	Transparency TransparencyConfig
+	PubSub       PubSubStorageConfig
+}
+
+type GCSStorageConfig struct {
+	Bucket string
+}
+
+type OCIStorageConfig struct {
+	Repository string
+	Insecure   bool
+}
+
+type PubSubStorageConfig struct {
+	Provider string
+	Topic    string
+	Kafka    KafkaStorageConfig
+}
+
+type KafkaStorageConfig struct {
+	BootstrapServers string
+}
+
+type TransparencyConfig struct {
+	Enabled          bool
+	VerifyAnnotation bool
+	URL              string
+}
+
+const (
+	gcsBucketKey             = "storage.gcs.bucket"
+	ociRepositoryKey         = "storage.oci.repository"
+	ociRepositoryInsecureKey = "storage.oci.repository.insecure"
+
+	// PubSub - General
+	pubsubProvider = "storage.pubsub.provider"
+	pubsubTopic    = "storage.pubsub.topic"
+
+	// No config for PubSub - In-Memory
+
+	// PubSub - Kafka
+	pubsubKafkaBootstrapServer = "storage.pubsub.kafka.bootstrap.servers"
+
+	transparencyEnabledKey = "transparency.enabled"
+	transparencyURLKey     = "transparency.url"
+)
+
+func defaultConfig() *Config {
+	return &Config{
+		Storage: StorageConfigs{
+			GCS: GCSStorageConfig{
+				Bucket: "URL",
+			},
+			OCI: OCIStorageConfig{
+				Repository: "URL",
+				Insecure:   false,
+			},
+			Transparency: TransparencyConfig{
+				URL: "https://rekor.sigstore.dev",
+			},
+		},
+	}
+}
+
+// NewConfigFromMap creates a Config from the supplied map
+func NewConfigFromMap(data map[string]string) (*Config, error) {
```

Comment:
hmm, can we just use json unmarshal here?

Reply:
Was thinking about this in the sense of configmap in k8s. We can iterate over this.

Reply:
removing config for now to figure out the better way to do this.